src/main/java/hudson/plugins/gradle/injection/MavenBuildScanInjection.java

```diff
@@ -63,29 +73,44 @@ private void injectMavenExtension(FilePath rootPath) {
             if (getGlobalEnvVar(GE_URL_VAR) != null) {
                 mavenOptsKeyValuePairs.add(asSystemProperty(GRADLE_ENTERPRISE_URL_PROPERTY_KEY, getGlobalEnvVar(GE_URL_VAR)));
             }
-            rootPath.act(new SetMavenOpts(mavenOptsKeyValuePairs));
+            appendMavenOpts(node, mavenOptsKeyValuePairs);
         } catch (IOException | InterruptedException e) {
             throw new IllegalStateException(e);
         }
     }
 
-    private void removeMavenExtension(FilePath rootPath) {
+    private void appendMavenOpts(Node node, List<String> mavenOptsKeyValuePairs) throws IOException, InterruptedException {
+        String mavenOpts = clearMavenOpts.removeSystemProperties(getMavenOpts(node)) + " " + String.join(" ", mavenOptsKeyValuePairs);
+        setMavenOpts(node, mavenOpts);
+    }
+
+    private void removeMavenExtension(Node node, FilePath rootPath) {
         try {
             deleteResourceFromAgent(GE_MVN_LIB_NAME, rootPath);
             deleteResourceFromAgent(CCUD_LIB_NAME, rootPath);
-            rootPath.act(new ClearMavenOpts(
-                    MAVEN_EXT_CLASS_PATH_PROPERTY_KEY,
-                    GRADLE_SCAN_UPLOAD_IN_BACKGROUND_PROPERTY_KEY,
-                    GRADLE_ENTERPRISE_ALLOW_UNTRUSTED_SERVER_PROPERTY_KEY,
-                    GRADLE_ENTERPRISE_URL_PROPERTY_KEY
-            ));
+            String mavenOpts = clearMavenOpts.removeSystemProperties(getMavenOpts(node));
+            setMavenOpts(node, mavenOpts);
```

Comment:
I think this object could do more. Getting and setting should also be part of it. I think it should look like this:

```java
class MavenOptsSetter {

    MavenOptsSetter(String... systemPropertyKeys) {
        //...
    }

    void appendIfMissing(Node node, String mavenOptsKeyValuePairs) {
        // get MAVEN_OPTS from node, check if we need to append and if so append them
    }

    void remove(Node node) {
        // get MAVEN_OPTS from the node, remove our keys and send the result back to the node
    }
}
```

Reply:
With this you could even write a unit test that makes sure we don't append multiple times and we don't destroy pre-existing values of `MAVEN_OPTS`.